core/src/advice/map.rs

```diff
@@ -17,32 +21,47 @@ use crate::utils::{ByteReader, ByteWriter, Deserializable, DeserializationError,
 /// associated with a given key onto the advice stack using `adv.push_mapval` instruction. The VM
 /// can also insert new values into the advice map during execution.
 #[derive(Debug, Clone, Default, PartialEq, Eq)]
-pub struct AdviceMap(BTreeMap<Word, Vec<Felt>>);
+pub struct AdviceMap(BTreeMap<Word, Arc<[Felt]>>);
 
 /// Pair representing a key-value entry in an [`AdviceMap`]
-type MapEntry = (Word, Vec<Felt>);
+type MapEntry = (Word, Arc<[Felt]>);
 
 impl AdviceMap {
     /// Creates a new advice map.
     pub fn new() -> Self {
-        Self(BTreeMap::<Word, Vec<Felt>>::new())
+        Self::default()
     }
 
     /// Returns the values associated with given key.
     pub fn get(&self, key: &Word) -> Option<&[Felt]> {
-        self.0.get(key).map(|v| v.as_slice())
+        self.0.get(key).map(|v| &**v)
+    }
+
+    /// Returns a pointer to the values associated with given key.
+    pub fn get_arc(&self, key: &Word) -> Option<&Arc<[Felt]>> {
+        self.0.get(key)
+    }
+
+    /// Returns true if the key has a corresponding value in the map.
+    pub fn contains_key(&self, key: &Word) -> bool {
+        self.0.contains_key(key)
     }
 
     /// Inserts a key value pair in the advice map and returns the inserted value.
-    pub fn insert(&mut self, key: Word, value: Vec<Felt>) -> Option<Vec<Felt>> {
-        self.0.insert(key, value)
+    pub fn insert(&mut self, key: Word, value: Vec<Felt>) -> Option<Arc<[Felt]>> {
```

Comment:
Should we make `value: Arc<[Felt]>`, so that if someone already has the value in this form, we can reuse the `Arc`? And if they have a `Vec`, they just need to `Arc::from()` it

Reply:
Made the value generic to take `Into<Arc<[T]>>`